### PR TITLE
Add Faster RCNN Resnet V1 FPN Keras feature extractor

### PR DESCRIPTION
**Pull request by syiming**
# Description

> :memo: Add faster RCNN Resnet V1 FPN Keras feature extractor
>  
> * Add faster RCNN Resnet {50, 101, 152} FPN Keras feature extractor.
> * Add unit test to ensure the size for proposal and box classifier features are correct.
> * Add model to model_builder.py
> * Need to further modify faster_rcnn_meta_arch.py to make the extractor work

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [ ] TensorFlow 2 migration
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] A new research paper code implementation
- [ ] Other (Specify)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: research/object_detection/builders/model_builder.py [hunk 1]
```diff
@@ -109,6 +110,12 @@
           frcnn_resnet_keras.FasterRCNNResnet152KerasFeatureExtractor,
       'faster_rcnn_inception_resnet_v2_keras':
       frcnn_inc_res_keras.FasterRCNNInceptionResnetV2KerasFeatureExtractor,
+      'fasret_rcnn_resnet50_fpn_keras':
+          frcnn_resnet_fpn_keras.FasterRCNNResnet50FPNKerasFeatureExtractor,
```
Comment (tombstone) [inline research/object_detection/builders/model_builder.py:114]: nit: To be consistent with other fpn feature extractors here, change FPN to Fpn

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            tf.keras.layers.Flatten(),
+            tf.keras.layers.Dense(units=1024, activation='relu'),
+            tf.keras.layers.Dense(units=1024, activation='relu')
+        ])
+        return feature_extractor_model
+
+
+class FasterRCNNResnet50FPNKerasFeatureExtractor(
+    FasterRCNNResnetV1FPNKerasFeatureExtractor):
+  """Faster RCNN with Resnet50 FPN feature extractor implementation."""
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:228]: drop `implementation`

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+        A 4-D float tensor with shape
+        [batch_size * self.max_num_proposals, 1024]
+        representing box classifier features for each proposal.
+    """
+    with tf.name_scope(name):
+      with tf.name_scope('ResnetV1FPN'):
+        feature_extractor_model = tf.keras.models.Sequential([
+            tf.keras.layers.Flatten(),
+            tf.keras.layers.Dense(units=1024, activation='relu'),
+            tf.keras.layers.Dense(units=1024, activation='relu')
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:221]: We will likely need a batchnorm between the two FC layers. We don't need to address it now, but please add a TODO comment here.

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+    faster_rcnn_meta_arch.FasterRCNNKerasFeatureExtractor):
+  """Faster RCNN Feature Extractor using Keras-based Resnet V1 FPN features."""
+
+  def __init__(self,
+               is_training,
+               resnet_v1_base_model,
+               resnet_v1_base_model_name,
+               first_stage_features_stride,
+               conv_hyperparams,
+               min_depth,
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:45]: drop `min_depth` and `depth_multipler`. No need to support these.

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+            include_top=False)
+        output_layers = _RESNET_MODEL_OUTPUT_LAYERS[self._resnet_v1_base_model_name]
+        outputs = [full_resnet_v1_model.get_layer(output_layer_name).output
+                   for output_layer_name in output_layers]
+        self.classification_backbone = tf.keras.Model(
+            inputs=full_resnet_v1_model.inputs,
+            outputs=outputs)
+        backbone_outputs = self.classification_backbone(full_resnet_v1_model.inputs)
+
+        # construct FPN feature generator
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:168]: remove the depth multiplier feature.

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+        for level in range(self._fpn_min_level, self._base_fpn_max_level + 1):
+          feature_block_list.append('block{}'.format(level - 1))
+        feature_block_map = dict(
+            list(zip(self._resnet_block_names, backbone_outputs)))
+        fpn_input_image_features = [
+            (feature_block, feature_block_map[feature_block])
+            for feature_block in feature_block_list]
+        fpn_features = self._fpn_features_generator(fpn_input_image_features)
+        features_maps = [fpn_feature for _, fpn_feature in fpn_features.items()]
+
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:192]: This is still missing coarse feature layers from self._base_fnp_max_level + 1 to self._fpn_max_level. Please see ssd_resnet_v1_fpn_keras_feature_extractor.py for reference.

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor_tf2_test.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        name='TestScope')(preprocessed_inputs)
+    features_shapes = [tf.shape(rpn_feature_map)
+                       for rpn_feature_map in rpn_feature_maps]
+
+    self.assertAllEqual(features_shapes[0].numpy(), [2, 56, 56, 256])
+    self.assertAllEqual(features_shapes[1].numpy(), [2, 28, 28, 256])
+    self.assertAllEqual(features_shapes[2].numpy(), [2, 14, 14, 256])
+    self.assertAllEqual(features_shapes[3].numpy(), [2, 7, 7, 256])
+
+  def test_extract_proposal_features_dies_with_incorrect_rank_inputs(self):
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor_tf2_test.py:82]: Drop this testcase. This is a bit too rigid and pollutes the test output.

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor_tf2_test.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        name='TestScope')
+    proposal_classifier_features = (
+        model(proposal_feature_maps))
+    features_shape = tf.shape(proposal_classifier_features)
+
+    self.assertAllEqual(features_shape.numpy(), [3, 1024])
+
+
+if __name__ == '__main__':
+  tf.enable_v2_behavior()
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor_tf2_test.py:104]: no need to enable. This is on by default.

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py (hunk 6 above)
Comment (tombstone) [reply research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:213]: Thanks for adding!

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+    faster_rcnn_meta_arch.FasterRCNNKerasFeatureExtractor):
+  """Faster RCNN Feature Extractor using Keras-based Resnet V1 FPN features."""
+
+  def __init__(self,
+               is_training,
+               resnet_v1_base_model,
+               resnet_v1_base_model_name,
+               first_stage_features_stride,
+               conv_hyperparams,
+               batch_norm_trainable=False,
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:45]: In a followup PR add fc_hyperparams and use this to configure batchnorm params for the layer that goes in between two FC layers.

File: research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+  """Faster RCNN with Resnet50 FPN feature extractor."""
+
+  def __init__(self,
+               is_training,
+               first_stage_features_stride=16,
+               conv_hyperparams=None,
+               batch_norm_trainable=False,
+               weight_decay=0.0,
+               fpn_min_level=2,
+               fpn_max_level=7,
```
Comment (tombstone) [inline research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py:259]: set the default max level to 6 as in the paper.
